<!DOCTYPE html>
<html xmlns="http://www.w3.org/1999/xhtml" lang="en" xml:lang="en"><head>

<meta charset="utf-8">
<meta name="generator" content="quarto-1.3.353">

<meta name="viewport" content="width=device-width, initial-scale=1.0, user-scalable=yes">

<meta name="author" content="John Zobolas">

<title>Survival modeling in mlr3 using Bayesian Additive Regression Trees (BART)</title>
<style>
code{white-space: pre-wrap;}
span.smallcaps{font-variant: small-caps;}
div.columns{display: flex; gap: min(4vw, 1.5em);}
div.column{flex: auto; overflow-x: auto;}
div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
ul.task-list{list-style: none;}
ul.task-list li input[type="checkbox"] {
  width: 0.8em;
  margin: 0 0.8em 0.2em -1em; /* quarto-specific, see https://github.com/quarto-dev/quarto-cli/issues/4556 */ 
  vertical-align: middle;
}
/* CSS for syntax highlighting */
pre > code.sourceCode { white-space: pre; position: relative; }
pre > code.sourceCode > span { display: inline-block; line-height: 1.25; }
pre > code.sourceCode > span:empty { height: 1.2em; }
.sourceCode { overflow: visible; }
code.sourceCode > span { color: inherit; text-decoration: inherit; }
div.sourceCode { margin: 1em 0; }
pre.sourceCode { margin: 0; }
@media screen {
div.sourceCode { overflow: auto; }
}
@media print {
pre > code.sourceCode { white-space: pre-wrap; }
pre > code.sourceCode > span { text-indent: -5em; padding-left: 5em; }
}
pre.numberSource code
  { counter-reset: source-line 0; }
pre.numberSource code > span
  { position: relative; left: -4em; counter-increment: source-line; }
pre.numberSource code > span > a:first-child::before
  { content: counter(source-line);
    position: relative; left: -1em; text-align: right; vertical-align: baseline;
    border: none; display: inline-block;
    -webkit-touch-callout: none; -webkit-user-select: none;
    -khtml-user-select: none; -moz-user-select: none;
    -ms-user-select: none; user-select: none;
    padding: 0 4px; width: 4em;
  }
pre.numberSource { margin-left: 3em;  padding-left: 4px; }
div.sourceCode
  {   }
@media screen {
pre > code.sourceCode > span > a:first-child::before { text-decoration: underline; }
}
/* CSS for citations */
div.csl-bib-body { }
div.csl-entry {
  clear: both;
}
.hanging-indent div.csl-entry {
  margin-left:2em;
  text-indent:-2em;
}
div.csl-left-margin {
  min-width:2em;
  float:left;
}
div.csl-right-inline {
  margin-left:2em;
  padding-left:1em;
}
div.csl-indent {
  margin-left: 2em;
}</style>


<script src="bart_post_files/libs/clipboard/clipboard.min.js"></script>
<script src="bart_post_files/libs/quarto-html/quarto.js"></script>
<script src="bart_post_files/libs/quarto-html/popper.min.js"></script>
<script src="bart_post_files/libs/quarto-html/tippy.umd.min.js"></script>
<script src="bart_post_files/libs/quarto-html/anchor.min.js"></script>
<link href="bart_post_files/libs/quarto-html/tippy.css" rel="stylesheet">
<link href="bart_post_files/libs/quarto-html/quarto-syntax-highlighting.css" rel="stylesheet" id="quarto-text-highlighting-styles">
<script src="bart_post_files/libs/bootstrap/bootstrap.min.js"></script>
<link href="bart_post_files/libs/bootstrap/bootstrap-icons.css" rel="stylesheet">
<link href="bart_post_files/libs/bootstrap/bootstrap.min.css" rel="stylesheet" id="quarto-bootstrap" data-mode="light">

  <script src="https://polyfill.io/v3/polyfill.min.js?features=es6"></script>
  <script src="https://cdn.jsdelivr.net/npm/mathjax@3/es5/tex-chtml-full.js" type="text/javascript"></script>

</head>

<body>

<div id="quarto-content" class="page-columns page-rows-contents page-layout-full">
<div id="quarto-margin-sidebar" class="sidebar margin-sidebar">
  <nav id="TOC" role="doc-toc" class="toc-active">
    <h2 id="toc-title">Table of contents</h2>
   
  <ul>
  <li><a href="#intro" id="toc-intro" class="nav-link active" data-scroll-target="#intro">Intro</a></li>
  <li><a href="#libraries" id="toc-libraries" class="nav-link" data-scroll-target="#libraries">Libraries</a></li>
  <li><a href="#data" id="toc-data" class="nav-link" data-scroll-target="#data">Data</a></li>
  <li><a href="#train-and-test" id="toc-train-and-test" class="nav-link" data-scroll-target="#train-and-test">Train and Test</a>
  <ul class="collapse">
  <li><a href="#distr" id="toc-distr" class="nav-link" data-scroll-target="#distr">distr</a></li>
  <li><a href="#crank" id="toc-crank" class="nav-link" data-scroll-target="#crank">crank</a></li>
  </ul></li>
  <li><a href="#feature-importance" id="toc-feature-importance" class="nav-link" data-scroll-target="#feature-importance">Feature importance</a></li>
  <li><a href="#mcmc-diagnostics" id="toc-mcmc-diagnostics" class="nav-link" data-scroll-target="#mcmc-diagnostics">MCMC Diagnostics</a></li>
  <li><a href="#performance-test-set" id="toc-performance-test-set" class="nav-link" data-scroll-target="#performance-test-set">Performance (test set)</a></li>
  <li><a href="#resampling" id="toc-resampling" class="nav-link" data-scroll-target="#resampling">Resampling</a></li>
  <li><a href="#uncertainty-quantification-in-survival-prediction" id="toc-uncertainty-quantification-in-survival-prediction" class="nav-link" data-scroll-target="#uncertainty-quantification-in-survival-prediction">Uncertainty Quantification in Survival Prediction</a></li>
  <li><a href="#partial-dependence-plot" id="toc-partial-dependence-plot" class="nav-link" data-scroll-target="#partial-dependence-plot">Partial Dependence Plot</a></li>
  <li><a href="#references" id="toc-references" class="nav-link" data-scroll-target="#references">References</a></li>
  </ul>
</nav>
</div>
<main class="content column-page-left" id="quarto-document-content">

<header id="title-block-header" class="quarto-title-block default">
<div class="quarto-title">
<h1 class="title">Survival modeling in mlr3 using Bayesian Additive Regression Trees (BART)</h1>
</div>



<div class="quarto-title-meta">

    <div>
    <div class="quarto-title-meta-heading">Author</div>
    <div class="quarto-title-meta-contents">
             <p>John Zobolas </p>
          </div>
  </div>
    
  
    
  </div>
  

</header>

<section id="intro" class="level2">
<h2 class="anchored" data-anchor-id="intro">Intro</h2>
<p>Here are some interesting reads regarding BART:</p>
<ul>
<li>The first BART paper <span class="citation" data-cites="Chipman2010">(<a href="#ref-Chipman2010" role="doc-biblioref">Chipman, George, and McCulloch 2010</a>)</span>.</li>
<li>The first implementation of BART for survival data <span class="citation" data-cites="Bonato2011">(<a href="#ref-Bonato2011" role="doc-biblioref">Bonato et al. 2011</a>)</span>. This includes fully parametric AFT and Weibull models and the semi-parametric CoxPH regression model.</li>
<li>The first non-parametric implementation of BART for survival data <span class="citation" data-cites="Sparapani2016">(<a href="#ref-Sparapani2016" role="doc-biblioref">R. A. Sparapani et al. 2016</a>)</span></li>
<li><code>BART</code> R package tutorial <span class="citation" data-cites="Sparapani2021">(<a href="#ref-Sparapani2021" role="doc-biblioref">R. Sparapani, Spanbauer, and McCulloch 2021</a>)</span></li>
</ul>
<p>We incorporated the survival <code>BART</code> model in <code>mlr3extralearners</code> and in this tutorial we will demonstrate how we can use packages like <code>mlr3</code>, <code>mlr3proba</code> and <code>distr6</code> to more easily manipulate the output predictions, evaluate the learner’s performance and graphically display them.</p>
</section>
<section id="libraries" class="level2">
<h2 class="anchored" data-anchor-id="libraries">Libraries</h2>
<div class="cell">
<div class="sourceCode cell-code" id="cb1"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb1-1"><a href="#cb1-1" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(mlr3extralearners)</span>
<span id="cb1-2"><a href="#cb1-2" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(mlr3pipelines)</span>
<span id="cb1-3"><a href="#cb1-3" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(mlr3proba)</span>
<span id="cb1-4"><a href="#cb1-4" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(distr6)</span>
<span id="cb1-5"><a href="#cb1-5" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(BART) <span class="co"># 2.9.4</span></span>
<span id="cb1-6"><a href="#cb1-6" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(dplyr)</span>
<span id="cb1-7"><a href="#cb1-7" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(tidyr)</span>
<span id="cb1-8"><a href="#cb1-8" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(tibble)</span>
<span id="cb1-9"><a href="#cb1-9" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(ggplot2)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</section>
<section id="data" class="level2">
<h2 class="anchored" data-anchor-id="data">Data</h2>
<p>We will use the Lung Cancer Dataset. We convert the <code>time</code> variable from days to months to ease the computational burden:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb2"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb2-1"><a href="#cb2-1" aria-hidden="true" tabindex="-1"></a>task_lung <span class="ot">=</span> <span class="fu">tsk</span>(<span class="st">'lung'</span>)</span>
<span id="cb2-2"><a href="#cb2-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb2-3"><a href="#cb2-3" aria-hidden="true" tabindex="-1"></a>d <span class="ot">=</span> task_lung<span class="sc">$</span><span class="fu">data</span>()</span>
<span id="cb2-4"><a href="#cb2-4" aria-hidden="true" tabindex="-1"></a><span class="co"># in case we want to select specific columns to keep</span></span>
<span id="cb2-5"><a href="#cb2-5" aria-hidden="true" tabindex="-1"></a><span class="co"># d = d[ ,colnames(d) %in% c("time", "status", "age", "sex", "ph.karno"), with = FALSE]</span></span>
<span id="cb2-6"><a href="#cb2-6" aria-hidden="true" tabindex="-1"></a>d<span class="sc">$</span>time <span class="ot">=</span> <span class="fu">ceiling</span>(d<span class="sc">$</span>time<span class="sc">/</span><span class="fl">30.44</span>)</span>
<span id="cb2-7"><a href="#cb2-7" aria-hidden="true" tabindex="-1"></a>task_lung <span class="ot">=</span> <span class="fu">as_task_surv</span>(d, <span class="at">time =</span> <span class="st">'time'</span>, <span class="at">event =</span> <span class="st">'status'</span>, <span class="at">id =</span> <span class="st">'lung'</span>)</span>
<span id="cb2-8"><a href="#cb2-8" aria-hidden="true" tabindex="-1"></a>task_lung<span class="sc">$</span>label <span class="ot">=</span> <span class="st">"Lung Cancer"</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div class="callout callout-style-default callout-note callout-titled">
<div class="callout-header d-flex align-content-center">
<div class="callout-icon-container">
<i class="callout-icon"></i>
</div>
<div class="callout-title-container flex-fill">
Note
</div>
</div>
<div class="callout-body-container callout-body">
<ol type="1">
<li>The original <code>BART</code> implementation supports categorical features (factors). This results in different importance scores per each dummy level which doesn’t work well with <code>mlr3</code>. So features of type <code>factor</code> or <code>character</code> are not allowed and we leave it to the user to encode them as they please.</li>
<li>The original <code>BART</code> implementation supports features with missing values. This is totally fine with <code>mlr3</code> as well! In this example, we impute the features to show good ML practice.</li>
</ol>
</div>
</div>
<p>In our lung dataset, we encode the <code>sex</code> feature and perform model-based imputation with the <code>rpart</code> regression learner:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb3"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb3-1"><a href="#cb3-1" aria-hidden="true" tabindex="-1"></a>po_encode <span class="ot">=</span> <span class="fu">po</span>(<span class="st">'encode'</span>, <span class="at">method =</span> <span class="st">'treatment'</span>)</span>
<span id="cb3-2"><a href="#cb3-2" aria-hidden="true" tabindex="-1"></a>po_impute <span class="ot">=</span> <span class="fu">po</span>(<span class="st">'imputelearner'</span>, <span class="fu">lrn</span>(<span class="st">'regr.rpart'</span>))</span>
<span id="cb3-3"><a href="#cb3-3" aria-hidden="true" tabindex="-1"></a>pre <span class="ot">=</span> po_encode <span class="sc">%&gt;&gt;%</span> po_impute</span>
<span id="cb3-4"><a href="#cb3-4" aria-hidden="true" tabindex="-1"></a>task <span class="ot">=</span> pre<span class="sc">$</span><span class="fu">train</span>(task_lung)[[<span class="dv">1</span>]]</span>
<span id="cb3-5"><a href="#cb3-5" aria-hidden="true" tabindex="-1"></a>task</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>&lt;TaskSurv:lung&gt; (228 x 10): Lung Cancer
* Target: time, status
* Properties: -
* Features (8):
  - int (7): age, inst, meal.cal, pat.karno, ph.ecog, ph.karno, wt.loss
  - dbl (1): sex</code></pre>
</div>
</div>
<p>No missing values in our data:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb5"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb5-1"><a href="#cb5-1" aria-hidden="true" tabindex="-1"></a>task<span class="sc">$</span><span class="fu">missings</span>()</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>     time    status       age       sex      inst  meal.cal pat.karno   ph.ecog 
        0         0         0         0         0         0         0         0 
 ph.karno   wt.loss 
        0         0 </code></pre>
</div>
</div>
<p>We partition the data to train and test sets:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb7"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb7-1"><a href="#cb7-1" aria-hidden="true" tabindex="-1"></a><span class="fu">set.seed</span>(<span class="dv">42</span>)</span>
<span id="cb7-2"><a href="#cb7-2" aria-hidden="true" tabindex="-1"></a>part <span class="ot">=</span> <span class="fu">partition</span>(task, <span class="at">ratio =</span> <span class="fl">0.9</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</section>
<section id="train-and-test" class="level2">
<h2 class="anchored" data-anchor-id="train-and-test">Train and Test</h2>
<p>We train the <code>BART</code> model and predict on the test set:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb8"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb8-1"><a href="#cb8-1" aria-hidden="true" tabindex="-1"></a><span class="co"># default `ndpost` value: 1000. We reduce it to 50 to speed up calculations in this tutorial</span></span>
<span id="cb8-2"><a href="#cb8-2" aria-hidden="true" tabindex="-1"></a>learner <span class="ot">=</span> <span class="fu">lrn</span>(<span class="st">"surv.bart"</span>, <span class="at">nskip =</span> <span class="dv">250</span>, <span class="at">ndpost =</span> <span class="dv">50</span>, <span class="at">keepevery =</span> <span class="dv">10</span>, <span class="at">mc.cores =</span> <span class="dv">10</span>)</span>
<span id="cb8-3"><a href="#cb8-3" aria-hidden="true" tabindex="-1"></a>learner<span class="sc">$</span><span class="fu">train</span>(task, <span class="at">row_ids =</span> part<span class="sc">$</span>train)</span>
<span id="cb8-4"><a href="#cb8-4" aria-hidden="true" tabindex="-1"></a>p <span class="ot">=</span> learner<span class="sc">$</span><span class="fu">predict</span>(task, <span class="at">row_ids =</span> part<span class="sc">$</span>test)</span>
<span id="cb8-5"><a href="#cb8-5" aria-hidden="true" tabindex="-1"></a>p</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>&lt;PredictionSurv&gt; for 23 observations:
    row_ids time status    crank     distr
          9    8   TRUE 66.19326 &lt;list[1]&gt;
         10    6   TRUE 98.43005 &lt;list[1]&gt;
         21   10   TRUE 54.82313 &lt;list[1]&gt;
---                                       
        160   13  FALSE 37.82089 &lt;list[1]&gt;
        163   10  FALSE 69.63534 &lt;list[1]&gt;
        194    8  FALSE 81.13678 &lt;list[1]&gt;</code></pre>
</div>
</div>
<p>See more details about <code>BART</code>’s parameters on the online <a href="https://mlr3extralearners.mlr-org.com/reference/mlr_learners_surv.bart.html">documentation</a>.</p>
<section id="distr" class="level3">
<h3 class="anchored" data-anchor-id="distr">distr</h3>
<p>What kind of object is the predicted <code>distr</code>?</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb10"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb10-1"><a href="#cb10-1" aria-hidden="true" tabindex="-1"></a>p<span class="sc">$</span>distr</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>Arrdist(23x31x50) </code></pre>
</div>
</div>
<div class="callout callout-style-default callout-tip callout-titled">
<div class="callout-header d-flex align-content-center">
<div class="callout-icon-container">
<i class="callout-icon"></i>
</div>
<div class="callout-title-container flex-fill">
Arrdist dimensions:
</div>
</div>
<div class="callout-body-container callout-body">
<ol type="1">
<li>Patients (observations)</li>
<li>Time points (months)</li>
<li>Number of posterior draws</li>
</ol>
</div>
</div>
<p>Actually the <code>$distr</code> is an active <a href="https://r6.r-lib.org/index.html">R6</a> field - this means that some computation is required to create it. What the prediction object actually stores internally is a 3d survival array (can be used directly with no performance overhead):</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb12"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb12-1"><a href="#cb12-1" aria-hidden="true" tabindex="-1"></a><span class="fu">dim</span>(p<span class="sc">$</span>data<span class="sc">$</span>distr)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>[1] 23 31 50</code></pre>
</div>
</div>
<p>This is a more easy-to-understand and manipulate form of the full posterior survival matrix prediction from the <code>BART</code> package (<span class="citation" data-cites="Sparapani2021">(<a href="#ref-Sparapani2021" role="doc-biblioref">R. Sparapani, Spanbauer, and McCulloch 2021</a>)</span>, pages 34-35).</p>
<div class="callout callout-style-default callout-warning callout-titled">
<div class="callout-header d-flex align-content-center">
<div class="callout-icon-container">
<i class="callout-icon"></i>
</div>
<div class="callout-title-container flex-fill">
Warning
</div>
</div>
<div class="callout-body-container callout-body">
<p>Though we have optimized with C++ code the way the <code>Arrdist</code> object is constructed, calling the <code>$distr</code> field can be computationally taxing if the product of the sizes of the 3 dimensions above <strong>exceeds ~1 million</strong>. In our case, <span class="math inline">\(23 \times 31 \times 50 = 35650\)</span> so the conversion to an <code>Arrdist</code> via <code>$distr</code> will certainly not create performance issues.</p>
</div>
</div>
<p>An example using the internal prediction data: get all the posterior probabilities of the 3rd patient in the test set, at 12 months (1 year):</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb14"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb14-1"><a href="#cb14-1" aria-hidden="true" tabindex="-1"></a>p<span class="sc">$</span>data<span class="sc">$</span>distr[<span class="dv">3</span>, <span class="dv">12</span>, ]</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code> [1] 0.26546909 0.27505937 0.21151435 0.46700513 0.26178380 0.24040003
 [7] 0.29946469 0.52357780 0.40833108 0.40367780 0.27027392 0.31781286
[13] 0.54151844 0.34460027 0.41826554 0.41866367 0.33694401 0.34511270
[19] 0.47244492 0.49423660 0.42069678 0.20095489 0.48696980 0.48409357
[25] 0.35649439 0.47969355 0.16355660 0.33728105 0.40245228 0.42418033
[31] 0.36336145 0.48181667 0.51858238 0.49635078 0.37238179 0.26694030
[37] 0.52219952 0.48992897 0.08572207 0.30306005 0.33881682 0.33463870
[43] 0.29102074 0.43176131 0.38554545 0.38053756 0.36808776 0.13772665
[49] 0.21898264 0.14552514</code></pre>
</div>
</div>
<p>Working with the <code>$distr</code> interface and <code>Arrdist</code> objects is very efficient as we will see later for predicting survival estimates.</p>
<div class="callout callout-style-default callout-tip callout-titled">
<div class="callout-header d-flex align-content-center">
<div class="callout-icon-container">
<i class="callout-icon"></i>
</div>
<div class="callout-title-container flex-fill">
Tip
</div>
</div>
<div class="callout-body-container callout-body">
<p>In survival analysis, <span class="math inline">\(S(t) = 1 - F(t)\)</span>, where <span class="math inline">\(S(t)\)</span> the survival function and <span class="math inline">\(F(t)\)</span> the cumulative distribution function (cdf). The latter can be interpreted as <code>risk</code> or probability of death up to time <span class="math inline">\(t\)</span>.</p>
<p>We can verify the above from the prediction object:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb16"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb16-1"><a href="#cb16-1" aria-hidden="true" tabindex="-1"></a>surv_array <span class="ot">=</span> <span class="dv">1</span> <span class="sc">-</span> distr6<span class="sc">::</span><span class="fu">gprm</span>(p<span class="sc">$</span>distr, <span class="st">"cdf"</span>) <span class="co"># 3d array</span></span>
<span id="cb16-2"><a href="#cb16-2" aria-hidden="true" tabindex="-1"></a>testthat<span class="sc">::</span><span class="fu">expect_equal</span>(p<span class="sc">$</span>data<span class="sc">$</span>distr, surv_array)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</div>
</div>
</section>
<section id="crank" class="level3">
<h3 class="anchored" data-anchor-id="crank">crank</h3>
<p><code>crank</code> is the <strong>expected mortality</strong> <span class="citation" data-cites="Sonabend2022">(<a href="#ref-Sonabend2022" role="doc-biblioref">Sonabend, Bender, and Vollmer 2022</a>)</span> which is the sum of the predicted cumulative hazard function (as is done in random survival forest models). Higher values denote larger risk. To calculate <code>crank</code>, we need a survival matrix. So we have to choose which 3rd dimension we should use from the predicted survival array. This is what the <code>which.curve</code> parameter of the <code>learner</code> does:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb17"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb17-1"><a href="#cb17-1" aria-hidden="true" tabindex="-1"></a>learner<span class="sc">$</span>param_set<span class="sc">$</span><span class="fu">get_values</span>()<span class="sc">$</span>which.curve</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>[1] 0.5</code></pre>
</div>
</div>
<p>The default value (<span class="math inline">\(0.5\)</span> quantile) is the <strong>median survival probability</strong>. It could be any other quantile (e.g.&nbsp;<span class="math inline">\(0.25\)</span>). Other possible values for <code>which.curve</code> are <code>mean</code> or a number denoting the exact posterior draw to extract (e.g.&nbsp;the last one, <code>which.curve = 50</code>).</p>
</section>
</section>
<section id="feature-importance" class="level2">
<h2 class="anchored" data-anchor-id="feature-importance">Feature importance</h2>
<p>Default score is the <strong>observed count of each feature</strong> in the trees (so the higher the score, the more important the feature):</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb19"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb19-1"><a href="#cb19-1" aria-hidden="true" tabindex="-1"></a>learner<span class="sc">$</span>param_set<span class="sc">$</span>values<span class="sc">$</span>importance</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>[1] "count"</code></pre>
</div>
<div class="sourceCode cell-code" id="cb21"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb21-1"><a href="#cb21-1" aria-hidden="true" tabindex="-1"></a>learner<span class="sc">$</span><span class="fu">importance</span>()</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>      sex  meal.cal      inst pat.karno  ph.karno   wt.loss       age   ph.ecog 
     7.84      7.46      7.08      6.76      6.60      6.46      5.48      5.42 </code></pre>
</div>
</div>
</section>
<section id="mcmc-diagnostics" class="level2">
<h2 class="anchored" data-anchor-id="mcmc-diagnostics">MCMC Diagnostics</h2>
<p><code>BART</code> uses internally MCMC (Markov Chain Monte Carlo) to sample from the posterior survival distribution. We need to check that MCMC has converged, meaning that the chains have reached a stationary distribution that approximates the true posterior survival distribution (otherwise the predictions may be inaccurate, misleading and unreliable).</p>
<p>We use Geweke’s convergence diagnostic test as it is implemented in the <code>BART</code> R package. We choose 10 random patients from the train set to evaluate the MCMC convergence.</p>
<div class="cell" data-layout-align="center">
<div class="sourceCode cell-code" id="cb23"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb23-1"><a href="#cb23-1" aria-hidden="true" tabindex="-1"></a><span class="co"># predictions on the train set</span></span>
<span id="cb23-2"><a href="#cb23-2" aria-hidden="true" tabindex="-1"></a>p_train <span class="ot">=</span> learner<span class="sc">$</span><span class="fu">predict</span>(task, <span class="at">row_ids =</span> part<span class="sc">$</span>train)</span>
<span id="cb23-3"><a href="#cb23-3" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb23-4"><a href="#cb23-4" aria-hidden="true" tabindex="-1"></a>z_list <span class="ot">=</span> <span class="fu">list</span>()</span>
<span id="cb23-5"><a href="#cb23-5" aria-hidden="true" tabindex="-1"></a><span class="co"># choose 10 patients from the train set randomly</span></span>
<span id="cb23-6"><a href="#cb23-6" aria-hidden="true" tabindex="-1"></a><span class="cf">for</span> (patient_id <span class="cf">in</span> <span class="fu">sample</span>(<span class="fu">length</span>(part<span class="sc">$</span>train), <span class="dv">10</span>)) {</span>
<span id="cb23-7"><a href="#cb23-7" aria-hidden="true" tabindex="-1"></a>  <span class="co"># matrix with columns =&gt; time points and rows =&gt; posterior draws</span></span>
<span id="cb23-8"><a href="#cb23-8" aria-hidden="true" tabindex="-1"></a>  post_surv <span class="ot">=</span> <span class="dv">1</span> <span class="sc">-</span> <span class="fu">t</span>(distr6<span class="sc">::</span><span class="fu">gprm</span>(p_train<span class="sc">$</span>distr[patient_id], <span class="st">"cdf"</span>)[<span class="dv">1</span>,,])</span>
<span id="cb23-9"><a href="#cb23-9" aria-hidden="true" tabindex="-1"></a>  z_list[[patient_id]] <span class="ot">=</span> BART<span class="sc">::</span><span class="fu">gewekediag</span>(post_surv)<span class="sc">$</span>z <span class="co"># get the z-scores</span></span>
<span id="cb23-10"><a href="#cb23-10" aria-hidden="true" tabindex="-1"></a>}</span>
<span id="cb23-11"><a href="#cb23-11" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb23-12"><a href="#cb23-12" aria-hidden="true" tabindex="-1"></a><span class="co"># plot the z scores vs time for all patients</span></span>
<span id="cb23-13"><a href="#cb23-13" aria-hidden="true" tabindex="-1"></a>dplyr<span class="sc">::</span><span class="fu">bind_rows</span>(z_list) <span class="sc">%&gt;%</span> </span>
<span id="cb23-14"><a href="#cb23-14" aria-hidden="true" tabindex="-1"></a>  tidyr<span class="sc">::</span><span class="fu">pivot_longer</span>(<span class="at">cols =</span> <span class="fu">everything</span>()) <span class="sc">%&gt;%</span></span>
<span id="cb23-15"><a href="#cb23-15" aria-hidden="true" tabindex="-1"></a>  <span class="fu">mutate</span>(<span class="at">name =</span> <span class="fu">as.numeric</span>(name)) <span class="sc">%&gt;%</span></span>
<span id="cb23-16"><a href="#cb23-16" aria-hidden="true" tabindex="-1"></a>  <span class="fu">ggplot</span>(<span class="fu">aes</span>(<span class="at">x =</span> name, <span class="at">y =</span> value)) <span class="sc">+</span> </span>
<span id="cb23-17"><a href="#cb23-17" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_point</span>() <span class="sc">+</span></span>
<span id="cb23-18"><a href="#cb23-18" aria-hidden="true" tabindex="-1"></a>  <span class="fu">labs</span>(<span class="at">x =</span> <span class="st">"Time (months)"</span>, <span class="at">y =</span> <span class="st">"Z-scores"</span>) <span class="sc">+</span></span>
<span id="cb23-19"><a href="#cb23-19" aria-hidden="true" tabindex="-1"></a>  <span class="co"># add critical values for a = 0.05</span></span>
<span id="cb23-20"><a href="#cb23-20" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_hline</span>(<span class="at">yintercept =</span> <span class="fl">1.96</span>, <span class="at">linetype =</span> <span class="st">'dashed'</span>, <span class="at">color =</span> <span class="st">"red"</span>) <span class="sc">+</span> </span>
<span id="cb23-21"><a href="#cb23-21" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_hline</span>(<span class="at">yintercept =</span> <span class="sc">-</span><span class="fl">1.96</span>, <span class="at">linetype =</span> <span class="st">'dashed'</span>, <span class="at">color =</span> <span class="st">"red"</span>) <span class="sc">+</span> </span>
<span id="cb23-22"><a href="#cb23-22" aria-hidden="true" tabindex="-1"></a>  <span class="fu">theme_bw</span>(<span class="at">base_size =</span> <span class="dv">14</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<div class="quarto-figure quarto-figure-center">
<figure class="figure">
<p><img src="bart_post_files/figure-html/diag-mcmc-1.png" class="img-fluid figure-img" width="672"></p>
<figcaption class="figure-caption">Geweke plot for MCMC diagnostics. Z-scores for the difference in the mean survival prediction between the first 10% and last 50% part of a Markov chain. The predictions are taken from 10 random patients in the train set. Red lines indicate the a = 0.05 critical line. Only a few z-scores exceed the 95% limits so we conclude that convergence has been attained.</figcaption>
</figure>
</div>
</div>
</div>
</section>
<section id="performance-test-set" class="level2">
<h2 class="anchored" data-anchor-id="performance-test-set">Performance (test set)</h2>
<p>We will use the following survival metrics:</p>
<ol type="1">
<li>Integrated Brier Score (requires a survival distribution prediction - <code>distr</code>)</li>
<li>Right-Censored Log loss (requires a survival distribution prediction - <code>distr</code>)</li>
<li>Uno’s C-index (requires a continuous ranking score prediction - <code>crank</code>)</li>
</ol>
<p>For the first two measures we will use the ERV (<strong>Explained Residual Variation</strong>) version, which standardizes the scores against a Kaplan-Meier (KM) baseline <span class="citation" data-cites="Sonabend2022a">(<a href="#ref-Sonabend2022a" role="doc-biblioref">Sonabend et al. 2022</a>)</span>. This means that values close to <span class="math inline">\(0\)</span> represent performance similar to a KM model, negative values denote worse performance than KM and <span class="math inline">\(1\)</span> is the absolute best possible score.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb24"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb24-1"><a href="#cb24-1" aria-hidden="true" tabindex="-1"></a>measures <span class="ot">=</span> <span class="fu">list</span>(</span>
<span id="cb24-2"><a href="#cb24-2" aria-hidden="true" tabindex="-1"></a>  <span class="fu">msr</span>(<span class="st">"surv.graf"</span>, <span class="at">ERV =</span> <span class="cn">TRUE</span>),</span>
<span id="cb24-3"><a href="#cb24-3" aria-hidden="true" tabindex="-1"></a>  <span class="fu">msr</span>(<span class="st">"surv.rcll"</span>, <span class="at">ERV =</span> <span class="cn">TRUE</span>),</span>
<span id="cb24-4"><a href="#cb24-4" aria-hidden="true" tabindex="-1"></a>  <span class="fu">msr</span>(<span class="st">"surv.cindex"</span>, <span class="at">weight_meth =</span> <span class="st">"G2"</span>, <span class="at">id =</span> <span class="st">"surv.cindex.uno"</span>)</span>
<span id="cb24-5"><a href="#cb24-5" aria-hidden="true" tabindex="-1"></a>)</span>
<span id="cb24-6"><a href="#cb24-6" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb24-7"><a href="#cb24-7" aria-hidden="true" tabindex="-1"></a><span class="cf">for</span> (measure <span class="cf">in</span> measures) {</span>
<span id="cb24-8"><a href="#cb24-8" aria-hidden="true" tabindex="-1"></a>  <span class="fu">print</span>(p<span class="sc">$</span><span class="fu">score</span>(measure, <span class="at">task =</span> task, <span class="at">train_set =</span> part<span class="sc">$</span>train))</span>
<span id="cb24-9"><a href="#cb24-9" aria-hidden="true" tabindex="-1"></a>}</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>  surv.graf 
-0.09950096 
  surv.rcll 
-0.02622117 
surv.cindex.uno 
       0.551951 </code></pre>
</div>
</div>
<div class="callout callout-style-default callout-note callout-titled">
<div class="callout-header d-flex align-content-center">
<div class="callout-icon-container">
<i class="callout-icon"></i>
</div>
<div class="callout-title-container flex-fill">
Note
</div>
</div>
<div class="callout-body-container callout-body">
<p>All metrics use by default the <strong>median survival distribution</strong> from the 3d array, no matter what is the <code>which.curve</code> argument during the learner’s construction.</p>
</div>
</div>
</section>
<section id="resampling" class="level2">
<h2 class="anchored" data-anchor-id="resampling">Resampling</h2>
<p>Performing resampling with the <code>BART</code> learner is very easy using <code>mlr3</code>.</p>
<p>We first stratify the data by <code>status</code>, so that in each resampling the proportion of censored vs un-censored patients remains the same:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb26"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb26-1"><a href="#cb26-1" aria-hidden="true" tabindex="-1"></a>task<span class="sc">$</span>col_roles<span class="sc">$</span>stratum <span class="ot">=</span> <span class="st">'status'</span></span>
<span id="cb26-2"><a href="#cb26-2" aria-hidden="true" tabindex="-1"></a>task<span class="sc">$</span>strata</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>     N                row_id
1: 165       1,2,4,5,7,8,...
2:  63  3, 6,38,68,71,83,...</code></pre>
</div>
</div>
<div class="cell" data-hash="bart_post_cache/html/unnamed-chunk-16_162ae8c615d482a93928602e9dab3bf6">
<div class="sourceCode cell-code" id="cb28"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb28-1"><a href="#cb28-1" aria-hidden="true" tabindex="-1"></a>rr <span class="ot">=</span> <span class="fu">resample</span>(task, learner, <span class="at">resampling =</span> <span class="fu">rsmp</span>(<span class="st">"cv"</span>, <span class="at">folds =</span> <span class="dv">5</span>), <span class="at">store_backends =</span> <span class="cn">TRUE</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>INFO  [13:00:29.593] [mlr3] Applying learner 'surv.bart' on task 'lung' (iter 1/5)
INFO  [13:00:31.355] [mlr3] Applying learner 'surv.bart' on task 'lung' (iter 2/5)
INFO  [13:00:33.035] [mlr3] Applying learner 'surv.bart' on task 'lung' (iter 3/5)
INFO  [13:00:35.067] [mlr3] Applying learner 'surv.bart' on task 'lung' (iter 4/5)
INFO  [13:00:36.691] [mlr3] Applying learner 'surv.bart' on task 'lung' (iter 5/5)</code></pre>
</div>
</div>
<p>No errors or warnings:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb30"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb30-1"><a href="#cb30-1" aria-hidden="true" tabindex="-1"></a>rr<span class="sc">$</span>errors</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>Empty data.table (0 rows and 2 cols): iteration,msg</code></pre>
</div>
<div class="sourceCode cell-code" id="cb32"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb32-1"><a href="#cb32-1" aria-hidden="true" tabindex="-1"></a>rr<span class="sc">$</span>warnings</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>Empty data.table (0 rows and 2 cols): iteration,msg</code></pre>
</div>
</div>
<p>Performance in each fold:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb34"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb34-1"><a href="#cb34-1" aria-hidden="true" tabindex="-1"></a>rr<span class="sc">$</span><span class="fu">score</span>(measures)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>   task_id learner_id resampling_id iteration    surv.graf    surv.rcll
1:    lung  surv.bart            cv         1 -0.312614598 -0.102013166
2:    lung  surv.bart            cv         2 -0.103181391 -0.009579343
3:    lung  surv.bart            cv         3  0.001448263  0.338851363
4:    lung  surv.bart            cv         4 -0.044161171  0.003691073
5:    lung  surv.bart            cv         5 -0.043129352  0.157902047
   surv.cindex.uno
1:       0.5869665
2:       0.5502903
3:       0.6178001
4:       0.6157215
5:       0.5688389
Hidden columns: task, learner, resampling, prediction</code></pre>
</div>
</div>
<p>Mean cross-validation performance:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb36"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb36-1"><a href="#cb36-1" aria-hidden="true" tabindex="-1"></a>rr<span class="sc">$</span><span class="fu">aggregate</span>(measures)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>      surv.graf       surv.rcll surv.cindex.uno 
     -0.1003276       0.0777704       0.5879235 </code></pre>
</div>
</div>
</section>
<section id="uncertainty-quantification-in-survival-prediction" class="level2">
<h2 class="anchored" data-anchor-id="uncertainty-quantification-in-survival-prediction">Uncertainty Quantification in Survival Prediction</h2>
<p>We will choose two patients from the test set and plot their survival prediction posterior estimates.</p>
<p>Let’s choose the patients with the worst and the best survival time:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb38"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb38-1"><a href="#cb38-1" aria-hidden="true" tabindex="-1"></a>death_times <span class="ot">=</span> p<span class="sc">$</span>truth[,<span class="dv">1</span>]</span>
<span id="cb38-2"><a href="#cb38-2" aria-hidden="true" tabindex="-1"></a><span class="fu">sort</span>(death_times)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code> [1]  3  5  5  6  6  6  7  8  8  8  8 10 10 10 12 12 12 13 15 16 17 18 27</code></pre>
</div>
<div class="sourceCode cell-code" id="cb40"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb40-1"><a href="#cb40-1" aria-hidden="true" tabindex="-1"></a>worst_indx <span class="ot">=</span> <span class="fu">which</span>(death_times <span class="sc">==</span> <span class="fu">min</span>(death_times))[<span class="dv">1</span>] <span class="co"># died first</span></span>
<span id="cb40-2"><a href="#cb40-2" aria-hidden="true" tabindex="-1"></a>best_indx  <span class="ot">=</span> <span class="fu">which</span>(death_times <span class="sc">==</span> <span class="fu">max</span>(death_times))[<span class="dv">1</span>] <span class="co"># died last</span></span>
<span id="cb40-3"><a href="#cb40-3" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb40-4"><a href="#cb40-4" aria-hidden="true" tabindex="-1"></a>patient_ids <span class="ot">=</span> <span class="fu">c</span>(worst_indx, best_indx)</span>
<span id="cb40-5"><a href="#cb40-5" aria-hidden="true" tabindex="-1"></a>patient_ids <span class="co"># which patient IDs</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>[1]  5 18</code></pre>
</div>
<div class="sourceCode cell-code" id="cb42"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb42-1"><a href="#cb42-1" aria-hidden="true" tabindex="-1"></a>death_times <span class="ot">=</span> death_times[patient_ids]</span>
<span id="cb42-2"><a href="#cb42-2" aria-hidden="true" tabindex="-1"></a>death_times <span class="co"># 1st is worst, 2nd is best</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>[1]  3 27</code></pre>
</div>
</div>
<p>Subset <code>Arrdist</code> to only the above 2 patients:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb44"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb44-1"><a href="#cb44-1" aria-hidden="true" tabindex="-1"></a>arrd <span class="ot">=</span> p<span class="sc">$</span>distr[patient_ids]</span>
<span id="cb44-2"><a href="#cb44-2" aria-hidden="true" tabindex="-1"></a>arrd</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>Arrdist(2x31x50) </code></pre>
</div>
</div>
<p>We choose time points (in months) for the survival estimates:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb46"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb46-1"><a href="#cb46-1" aria-hidden="true" tabindex="-1"></a>months <span class="ot">=</span> <span class="fu">seq</span>(<span class="dv">1</span>, <span class="dv">36</span>) <span class="co"># 1 month - 3 years</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>We use the <code>$distr</code> interface and the <code>$survival</code> property to get survival probabilities from an <code>Arrdist</code> object as well as the <strong>quantile credible intervals</strong> (CIs). The median survival probabilities can be extracted as follows:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb47"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb47-1"><a href="#cb47-1" aria-hidden="true" tabindex="-1"></a>med <span class="ot">=</span> arrd<span class="sc">$</span><span class="fu">survival</span>(months) <span class="co"># 'med' for median</span></span>
<span id="cb47-2"><a href="#cb47-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb47-3"><a href="#cb47-3" aria-hidden="true" tabindex="-1"></a><span class="fu">colnames</span>(med) <span class="ot">=</span> <span class="fu">paste0</span>(patient_ids, <span class="st">"_med"</span>)</span>
<span id="cb47-4"><a href="#cb47-4" aria-hidden="true" tabindex="-1"></a>med <span class="ot">=</span> <span class="fu">as_tibble</span>(med) <span class="sc">%&gt;%</span> <span class="fu">add_column</span>(<span class="at">month =</span> months)</span>
<span id="cb47-5"><a href="#cb47-5" aria-hidden="true" tabindex="-1"></a><span class="fu">head</span>(med)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code># A tibble: 6 × 3
  `5_med` `18_med` month
    &lt;dbl&gt;    &lt;dbl&gt; &lt;int&gt;
1   0.874    0.981     1
2   0.767    0.962     2
3   0.670    0.945     3
4   0.569    0.927     4
5   0.465    0.901     5
6   0.366    0.869     6</code></pre>
</div>
</div>
<p>We can briefly verify model’s predictions: 1st patient survival probabilities on any month are lower (worst) compared to the 2nd patient.</p>
<p>Note that subsetting an <code>Arrdist</code> (3d array) creates a <code>Matdist</code> (2d matrix), for example we can explicitly get the median survival probabilities:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb49"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb49-1"><a href="#cb49-1" aria-hidden="true" tabindex="-1"></a>matd_median <span class="ot">=</span> arrd[, <span class="fl">0.5</span>] <span class="co"># median</span></span>
<span id="cb49-2"><a href="#cb49-2" aria-hidden="true" tabindex="-1"></a><span class="fu">head</span>(matd_median<span class="sc">$</span><span class="fu">survival</span>(months)) <span class="co"># same as with `arrd`</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>       [,1]      [,2]
1 0.8741127 0.9808363
2 0.7670382 0.9621618
3 0.6701276 0.9450867
4 0.5688809 0.9272284
5 0.4647686 0.9007042
6 0.3660939 0.8687270</code></pre>
</div>
</div>
<p>Using the <code>mean</code> posterior survival probabilities or the ones from the last posterior draw is also possible and can be done as follows:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb51"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb51-1"><a href="#cb51-1" aria-hidden="true" tabindex="-1"></a>matd_mean <span class="ot">=</span> arrd[, <span class="st">"mean"</span>] <span class="co"># mean (if needed)</span></span>
<span id="cb51-2"><a href="#cb51-2" aria-hidden="true" tabindex="-1"></a><span class="fu">head</span>(matd_mean<span class="sc">$</span><span class="fu">survival</span>(months))</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>       [,1]      [,2]
1 0.8652006 0.9748463
2 0.7533538 0.9521817
3 0.6560050 0.9293229
4 0.5623555 0.9051549
5 0.4750038 0.8758896
6 0.3815333 0.8360373</code></pre>
</div>
<div class="sourceCode cell-code" id="cb53"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb53-1"><a href="#cb53-1" aria-hidden="true" tabindex="-1"></a>matd_50draw <span class="ot">=</span> arrd[, <span class="dv">50</span>] <span class="co"># the 50th posterior draw</span></span>
<span id="cb53-2"><a href="#cb53-2" aria-hidden="true" tabindex="-1"></a><span class="fu">head</span>(matd_50draw<span class="sc">$</span><span class="fu">survival</span>(months))</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>       [,1]      [,2]
1 0.9178342 0.9920982
2 0.8424195 0.9842589
3 0.7732014 0.9764815
4 0.7096707 0.9687656
5 0.6029119 0.9495583
6 0.5122132 0.9307318</code></pre>
</div>
</div>
<p>To get the CIs we will subset the <code>Arrdist</code> using a quantile number (0-1), which extracts a <code>Matdist</code> based on the cdf. The survival function is 1 - cdf, so low and upper bounds are reversed:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb55"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb55-1"><a href="#cb55-1" aria-hidden="true" tabindex="-1"></a>low  <span class="ot">=</span> arrd[, <span class="fl">0.975</span>]<span class="sc">$</span><span class="fu">survival</span>(months) <span class="co"># 2.5% bound</span></span>
<span id="cb55-2"><a href="#cb55-2" aria-hidden="true" tabindex="-1"></a>high <span class="ot">=</span> arrd[, <span class="fl">0.025</span>]<span class="sc">$</span><span class="fu">survival</span>(months) <span class="co"># 97.5% bound</span></span>
<span id="cb55-3"><a href="#cb55-3" aria-hidden="true" tabindex="-1"></a><span class="fu">colnames</span>(low)  <span class="ot">=</span> <span class="fu">paste0</span>(patient_ids, <span class="st">"_low"</span>)</span>
<span id="cb55-4"><a href="#cb55-4" aria-hidden="true" tabindex="-1"></a><span class="fu">colnames</span>(high) <span class="ot">=</span> <span class="fu">paste0</span>(patient_ids, <span class="st">"_high"</span>)</span>
<span id="cb55-5"><a href="#cb55-5" aria-hidden="true" tabindex="-1"></a>low  <span class="ot">=</span> <span class="fu">as_tibble</span>(low)</span>
<span id="cb55-6"><a href="#cb55-6" aria-hidden="true" tabindex="-1"></a>high <span class="ot">=</span> <span class="fu">as_tibble</span>(high)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>The median posterior survival probabilities for the two patient of interest and the corresponding CI bounds in a tidy format are:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb56"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb56-1"><a href="#cb56-1" aria-hidden="true" tabindex="-1"></a>surv_tbl <span class="ot">=</span></span>
<span id="cb56-2"><a href="#cb56-2" aria-hidden="true" tabindex="-1"></a>  <span class="fu">bind_cols</span>(low, med, high) <span class="sc">%&gt;%</span></span>
<span id="cb56-3"><a href="#cb56-3" aria-hidden="true" tabindex="-1"></a>  <span class="fu">pivot_longer</span>(<span class="at">cols =</span> <span class="sc">!</span>month, <span class="at">values_to =</span> <span class="st">"surv"</span>,</span>
<span id="cb56-4"><a href="#cb56-4" aria-hidden="true" tabindex="-1"></a>    <span class="at">names_to =</span> <span class="fu">c</span>(<span class="st">"patient_id"</span>, <span class="st">".value"</span>), <span class="at">names_sep =</span> <span class="st">"_"</span>) <span class="sc">%&gt;%</span></span>
<span id="cb56-5"><a href="#cb56-5" aria-hidden="true" tabindex="-1"></a>  <span class="fu">relocate</span>(patient_id)</span>
<span id="cb56-6"><a href="#cb56-6" aria-hidden="true" tabindex="-1"></a>surv_tbl</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code># A tibble: 72 × 5
   patient_id month   low   med  high
   &lt;chr&gt;      &lt;int&gt; &lt;dbl&gt; &lt;dbl&gt; &lt;dbl&gt;
 1 5              1 0.713 0.874 0.953
 2 18             1 0.929 0.981 0.996
 3 5              2 0.508 0.767 0.903
 4 18             2 0.863 0.962 0.991
 5 5              3 0.362 0.670 0.855
 6 18             3 0.801 0.945 0.985
 7 5              4 0.244 0.569 0.804
 8 18             4 0.734 0.927 0.977
 9 5              5 0.146 0.465 0.748
10 18             5 0.654 0.901 0.969
# ℹ 62 more rows</code></pre>
</div>
</div>
<p>We draw survival curves with the uncertainty for the survival probability quantified:</p>
<div class="cell" data-layout-align="center">
<div class="sourceCode cell-code" id="cb58"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb58-1"><a href="#cb58-1" aria-hidden="true" tabindex="-1"></a>my_colors <span class="ot">=</span> <span class="fu">c</span>(<span class="st">"#E41A1C"</span>, <span class="st">"#4DAF4A"</span>)</span>
<span id="cb58-2"><a href="#cb58-2" aria-hidden="true" tabindex="-1"></a><span class="fu">names</span>(my_colors) <span class="ot">=</span> patient_ids</span>
<span id="cb58-3"><a href="#cb58-3" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb58-4"><a href="#cb58-4" aria-hidden="true" tabindex="-1"></a>surv_tbl <span class="sc">%&gt;%</span></span>
<span id="cb58-5"><a href="#cb58-5" aria-hidden="true" tabindex="-1"></a>  <span class="fu">ggplot</span>(<span class="fu">aes</span>(<span class="at">x =</span> month, <span class="at">y =</span> med)) <span class="sc">+</span></span>
<span id="cb58-6"><a href="#cb58-6" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_step</span>(<span class="fu">aes</span>(<span class="at">color =</span> patient_id), <span class="at">linewidth =</span> <span class="dv">1</span>) <span class="sc">+</span></span>
<span id="cb58-7"><a href="#cb58-7" aria-hidden="true" tabindex="-1"></a>  <span class="fu">xlab</span>(<span class="st">'Time (Months)'</span>) <span class="sc">+</span></span>
<span id="cb58-8"><a href="#cb58-8" aria-hidden="true" tabindex="-1"></a>  <span class="fu">ylab</span>(<span class="st">'Survival Probability'</span>) <span class="sc">+</span></span>
<span id="cb58-9"><a href="#cb58-9" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_ribbon</span>(<span class="fu">aes</span>(<span class="at">ymin =</span> low, <span class="at">ymax =</span> high, <span class="at">fill =</span> patient_id), </span>
<span id="cb58-10"><a href="#cb58-10" aria-hidden="true" tabindex="-1"></a>    <span class="at">alpha =</span> <span class="fl">0.3</span>, <span class="at">show.legend =</span> F) <span class="sc">+</span></span>
<span id="cb58-11"><a href="#cb58-11" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_vline</span>(<span class="at">xintercept =</span> death_times[<span class="dv">1</span>], <span class="at">linetype =</span> <span class="st">'dashed'</span>, <span class="at">color =</span> my_colors[<span class="dv">1</span>]) <span class="sc">+</span></span>
<span id="cb58-12"><a href="#cb58-12" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_vline</span>(<span class="at">xintercept =</span> death_times[<span class="dv">2</span>], <span class="at">linetype =</span> <span class="st">'dashed'</span>, <span class="at">color =</span> my_colors[<span class="dv">2</span>]) <span class="sc">+</span></span>
<span id="cb58-13"><a href="#cb58-13" aria-hidden="true" tabindex="-1"></a>  <span class="fu">theme_bw</span>(<span class="at">base_size =</span> <span class="dv">14</span>) <span class="sc">+</span></span>
<span id="cb58-14"><a href="#cb58-14" aria-hidden="true" tabindex="-1"></a>  <span class="fu">scale_color_manual</span>(<span class="at">values =</span> my_colors) <span class="sc">+</span></span>
<span id="cb58-15"><a href="#cb58-15" aria-hidden="true" tabindex="-1"></a>  <span class="fu">scale_fill_manual</span>(<span class="at">values =</span> my_colors) <span class="sc">+</span></span>
<span id="cb58-16"><a href="#cb58-16" aria-hidden="true" tabindex="-1"></a>  <span class="fu">guides</span>(<span class="at">color =</span> <span class="fu">guide_legend</span>(<span class="at">title =</span> <span class="st">"Patient ID"</span>))</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<div class="quarto-figure quarto-figure-center">
<figure class="figure">
<p><img src="bart_post_files/figure-html/surv-with-uq-1.png" class="img-fluid figure-img" width="672"></p>
<figcaption class="figure-caption">Uncertainty quantification for the survival prediction of two patients in the test set using 95% credible intervals. The two vertical lines correspond to the reported time of death (in months) for the two patients.</figcaption>
</figure>
</div>
</div>
</div>
</section>
<section id="partial-dependence-plot" class="level2">
<h2 class="anchored" data-anchor-id="partial-dependence-plot">Partial Dependence Plot</h2>
<p>We will use a Partial Dependence Plot (PDP) <span class="citation" data-cites="Friedman2001">(<a href="#ref-Friedman2001" role="doc-biblioref">Friedman 2001</a>)</span> to visualize how much different are males vs females in terms of their average survival predictions across time.</p>
<div class="callout callout-style-default callout-note callout-titled">
<div class="callout-header d-flex align-content-center">
<div class="callout-icon-container">
<i class="callout-icon"></i>
</div>
<div class="callout-title-container flex-fill">
Note
</div>
</div>
<div class="callout-body-container callout-body">
<p>PDPs assume that features are independent. In our case we need to check that <code>sex</code> doesn’t correlate with any of the other features used for training the <code>BART</code> learner. Since <code>sex</code> is a categorical feature, we fit a linear model using as target variable every other feature in the data (<span class="math inline">\(lm(feature \sim sex)\)</span>) and conduct an ANOVA (ANalysis Of VAriance) to get the variance explained or <span class="math inline">\(R^2\)</span>. The square root of that value is the correlation measure we want.</p>
</div>
</div>
<div class="cell">
<div class="sourceCode cell-code" id="cb59"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb59-1"><a href="#cb59-1" aria-hidden="true" tabindex="-1"></a><span class="co"># code from https://christophm.github.io/interpretable-ml-book/ale.html</span></span>
<span id="cb59-2"><a href="#cb59-2" aria-hidden="true" tabindex="-1"></a>mycor <span class="ot">=</span> <span class="cf">function</span>(cnames, data) {</span>
<span id="cb59-3"><a href="#cb59-3" aria-hidden="true" tabindex="-1"></a>  x.num <span class="ot">=</span> data[, cnames[<span class="dv">1</span>], with <span class="ot">=</span> <span class="cn">FALSE</span>][[<span class="dv">1</span>]]</span>
<span id="cb59-4"><a href="#cb59-4" aria-hidden="true" tabindex="-1"></a>  x.cat <span class="ot">=</span> data[, cnames[<span class="dv">2</span>], with <span class="ot">=</span> <span class="cn">FALSE</span>][[<span class="dv">1</span>]]</span>
<span id="cb59-5"><a href="#cb59-5" aria-hidden="true" tabindex="-1"></a>  <span class="co"># R^2 = Cor(X, Y)^2 in simple linear regression</span></span>
<span id="cb59-6"><a href="#cb59-6" aria-hidden="true" tabindex="-1"></a>  <span class="fu">sqrt</span>(<span class="fu">summary</span>(<span class="fu">lm</span>(x.num <span class="sc">~</span> x.cat))<span class="sc">$</span>r.squared)</span>
<span id="cb59-7"><a href="#cb59-7" aria-hidden="true" tabindex="-1"></a>}</span>
<span id="cb59-8"><a href="#cb59-8" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb59-9"><a href="#cb59-9" aria-hidden="true" tabindex="-1"></a>cnames <span class="ot">=</span> <span class="fu">c</span>(<span class="st">"sex"</span>)</span>
<span id="cb59-10"><a href="#cb59-10" aria-hidden="true" tabindex="-1"></a>combs <span class="ot">=</span> <span class="fu">expand.grid</span>(<span class="at">y =</span> <span class="fu">setdiff</span>(<span class="fu">colnames</span>(d), <span class="st">"sex"</span>), <span class="at">x =</span> cnames)</span>
<span id="cb59-11"><a href="#cb59-11" aria-hidden="true" tabindex="-1"></a>combs<span class="sc">$</span>cor <span class="ot">=</span> <span class="fu">apply</span>(combs, <span class="dv">1</span>, mycor, <span class="at">data =</span> task<span class="sc">$</span><span class="fu">data</span>()) <span class="co"># use the train set</span></span>
<span id="cb59-12"><a href="#cb59-12" aria-hidden="true" tabindex="-1"></a>combs</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>          y   x        cor
1      time sex 0.12941337
2    status sex 0.24343282
3       age sex 0.12216709
4      inst sex 0.07826337
5  meal.cal sex 0.18389545
6 pat.karno sex 0.04132443
7   ph.ecog sex 0.02564987
8  ph.karno sex 0.01702471
9   wt.loss sex 0.13431983</code></pre>
</div>
</div>
<p><code>sex</code> doesn’t correlate strongly with any other feature, so we can compute the PDP:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb61"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb61-1"><a href="#cb61-1" aria-hidden="true" tabindex="-1"></a><span class="co"># create two datasets: one with males and one with females</span></span>
<span id="cb61-2"><a href="#cb61-2" aria-hidden="true" tabindex="-1"></a><span class="co"># all other features remain the same (use train data, 205 patients)</span></span>
<span id="cb61-3"><a href="#cb61-3" aria-hidden="true" tabindex="-1"></a>d <span class="ot">=</span> task<span class="sc">$</span><span class="fu">data</span>(<span class="at">rows =</span> part<span class="sc">$</span>train) <span class="co"># `rows = part$test` to use the test set</span></span>
<span id="cb61-4"><a href="#cb61-4" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb61-5"><a href="#cb61-5" aria-hidden="true" tabindex="-1"></a>d<span class="sc">$</span>sex <span class="ot">=</span> <span class="dv">1</span></span>
<span id="cb61-6"><a href="#cb61-6" aria-hidden="true" tabindex="-1"></a>task_males <span class="ot">=</span> <span class="fu">as_task_surv</span>(d, <span class="at">time =</span> <span class="st">'time'</span>, <span class="at">event =</span> <span class="st">'status'</span>, <span class="at">id =</span> <span class="st">'lung-males'</span>)</span>
<span id="cb61-7"><a href="#cb61-7" aria-hidden="true" tabindex="-1"></a>d<span class="sc">$</span>sex <span class="ot">=</span> <span class="dv">0</span></span>
<span id="cb61-8"><a href="#cb61-8" aria-hidden="true" tabindex="-1"></a>task_females <span class="ot">=</span> <span class="fu">as_task_surv</span>(d, <span class="at">time =</span> <span class="st">'time'</span>, <span class="at">event =</span> <span class="st">'status'</span>, <span class="at">id =</span> <span class="st">'lung-females'</span>)</span>
<span id="cb61-9"><a href="#cb61-9" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb61-10"><a href="#cb61-10" aria-hidden="true" tabindex="-1"></a><span class="co"># make predictions</span></span>
<span id="cb61-11"><a href="#cb61-11" aria-hidden="true" tabindex="-1"></a>p_males   <span class="ot">=</span> learner<span class="sc">$</span><span class="fu">predict</span>(task_males)</span>
<span id="cb61-12"><a href="#cb61-12" aria-hidden="true" tabindex="-1"></a>p_females <span class="ot">=</span> learner<span class="sc">$</span><span class="fu">predict</span>(task_females)</span>
<span id="cb61-13"><a href="#cb61-13" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb61-14"><a href="#cb61-14" aria-hidden="true" tabindex="-1"></a><span class="co"># take the median posterior survival probability</span></span>
<span id="cb61-15"><a href="#cb61-15" aria-hidden="true" tabindex="-1"></a>surv_males   <span class="ot">=</span> p_males<span class="sc">$</span>distr<span class="sc">$</span><span class="fu">survival</span>(months) <span class="co"># patients x times</span></span>
<span id="cb61-16"><a href="#cb61-16" aria-hidden="true" tabindex="-1"></a>surv_females <span class="ot">=</span> p_females<span class="sc">$</span>distr<span class="sc">$</span><span class="fu">survival</span>(months) <span class="co"># patients x times</span></span>
<span id="cb61-17"><a href="#cb61-17" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb61-18"><a href="#cb61-18" aria-hidden="true" tabindex="-1"></a><span class="co"># tidy up data: average and quantiles across patients</span></span>
<span id="cb61-19"><a href="#cb61-19" aria-hidden="true" tabindex="-1"></a>data_males <span class="ot">=</span> </span>
<span id="cb61-20"><a href="#cb61-20" aria-hidden="true" tabindex="-1"></a>  <span class="fu">apply</span>(surv_males, <span class="dv">1</span>, <span class="cf">function</span>(row) {</span>
<span id="cb61-21"><a href="#cb61-21" aria-hidden="true" tabindex="-1"></a>    <span class="fu">tibble</span>(</span>
<span id="cb61-22"><a href="#cb61-22" aria-hidden="true" tabindex="-1"></a>      <span class="at">low =</span> <span class="fu">quantile</span>(row, <span class="at">probs =</span> <span class="fl">0.025</span>), </span>
<span id="cb61-23"><a href="#cb61-23" aria-hidden="true" tabindex="-1"></a>      <span class="at">avg =</span> <span class="fu">mean</span>(row), </span>
<span id="cb61-24"><a href="#cb61-24" aria-hidden="true" tabindex="-1"></a>      <span class="at">high =</span> <span class="fu">quantile</span>(row, <span class="at">probs =</span> <span class="fl">0.975</span>)</span>
<span id="cb61-25"><a href="#cb61-25" aria-hidden="true" tabindex="-1"></a>    )</span>
<span id="cb61-26"><a href="#cb61-26" aria-hidden="true" tabindex="-1"></a>  }) <span class="sc">%&gt;%</span></span>
<span id="cb61-27"><a href="#cb61-27" aria-hidden="true" tabindex="-1"></a>  <span class="fu">bind_rows</span>() <span class="sc">%&gt;%</span> </span>
<span id="cb61-28"><a href="#cb61-28" aria-hidden="true" tabindex="-1"></a>  <span class="fu">add_column</span>(<span class="at">sex =</span> <span class="st">'male'</span>, <span class="at">month =</span> months, <span class="at">.before =</span> <span class="dv">1</span>)</span>
<span id="cb61-29"><a href="#cb61-29" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb61-30"><a href="#cb61-30" aria-hidden="true" tabindex="-1"></a>data_females <span class="ot">=</span> </span>
<span id="cb61-31"><a href="#cb61-31" aria-hidden="true" tabindex="-1"></a>  <span class="fu">apply</span>(surv_females, <span class="dv">1</span>, <span class="cf">function</span>(row) {</span>
<span id="cb61-32"><a href="#cb61-32" aria-hidden="true" tabindex="-1"></a>    <span class="fu">tibble</span>(</span>
<span id="cb61-33"><a href="#cb61-33" aria-hidden="true" tabindex="-1"></a>      <span class="at">low =</span> <span class="fu">quantile</span>(row, <span class="at">probs =</span> <span class="fl">0.025</span>), </span>
<span id="cb61-34"><a href="#cb61-34" aria-hidden="true" tabindex="-1"></a>      <span class="at">avg =</span> <span class="fu">mean</span>(row), </span>
<span id="cb61-35"><a href="#cb61-35" aria-hidden="true" tabindex="-1"></a>      <span class="at">high =</span> <span class="fu">quantile</span>(row, <span class="at">probs =</span> <span class="fl">0.975</span>)</span>
<span id="cb61-36"><a href="#cb61-36" aria-hidden="true" tabindex="-1"></a>    )</span>
<span id="cb61-37"><a href="#cb61-37" aria-hidden="true" tabindex="-1"></a>  }) <span class="sc">%&gt;%</span></span>
<span id="cb61-38"><a href="#cb61-38" aria-hidden="true" tabindex="-1"></a>  <span class="fu">bind_rows</span>() <span class="sc">%&gt;%</span> </span>
<span id="cb61-39"><a href="#cb61-39" aria-hidden="true" tabindex="-1"></a>  <span class="fu">add_column</span>(<span class="at">sex =</span> <span class="st">'female'</span>, <span class="at">month =</span> months, <span class="at">.before =</span> <span class="dv">1</span>)</span>
<span id="cb61-40"><a href="#cb61-40" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb61-41"><a href="#cb61-41" aria-hidden="true" tabindex="-1"></a>pdp_tbl <span class="ot">=</span> <span class="fu">bind_rows</span>(data_males, data_females)</span>
<span id="cb61-42"><a href="#cb61-42" aria-hidden="true" tabindex="-1"></a>pdp_tbl</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code># A tibble: 72 × 5
   sex   month    low   avg  high
   &lt;chr&gt; &lt;int&gt;  &lt;dbl&gt; &lt;dbl&gt; &lt;dbl&gt;
 1 male      1 0.836  0.942 0.981
 2 male      2 0.704  0.889 0.963
 3 male      3 0.587  0.839 0.943
 4 male      4 0.488  0.788 0.924
 5 male      5 0.392  0.732 0.897
 6 male      6 0.304  0.663 0.860
 7 male      7 0.234  0.601 0.829
 8 male      8 0.172  0.550 0.799
 9 male      9 0.130  0.503 0.766
10 male     10 0.0945 0.455 0.733
# ℹ 62 more rows</code></pre>
</div>
</div>
<div class="cell" data-layout-align="center">
<div class="sourceCode cell-code" id="cb63"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb63-1"><a href="#cb63-1" aria-hidden="true" tabindex="-1"></a>my_colors <span class="ot">=</span> <span class="fu">c</span>(<span class="st">"#E41A1C"</span>, <span class="st">"#4DAF4A"</span>)</span>
<span id="cb63-2"><a href="#cb63-2" aria-hidden="true" tabindex="-1"></a><span class="fu">names</span>(my_colors) <span class="ot">=</span> <span class="fu">c</span>(<span class="st">'male'</span>, <span class="st">'female'</span>)</span>
<span id="cb63-3"><a href="#cb63-3" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb63-4"><a href="#cb63-4" aria-hidden="true" tabindex="-1"></a>pdp_tbl <span class="sc">%&gt;%</span></span>
<span id="cb63-5"><a href="#cb63-5" aria-hidden="true" tabindex="-1"></a>  <span class="fu">ggplot</span>(<span class="fu">aes</span>(<span class="at">x =</span> month, <span class="at">y =</span> avg)) <span class="sc">+</span></span>
<span id="cb63-6"><a href="#cb63-6" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_step</span>(<span class="fu">aes</span>(<span class="at">color =</span> sex), <span class="at">linewidth =</span> <span class="dv">1</span>) <span class="sc">+</span></span>
<span id="cb63-7"><a href="#cb63-7" aria-hidden="true" tabindex="-1"></a>  <span class="fu">xlab</span>(<span class="st">'Time (Months)'</span>) <span class="sc">+</span></span>
<span id="cb63-8"><a href="#cb63-8" aria-hidden="true" tabindex="-1"></a>  <span class="fu">ylab</span>(<span class="st">'Survival Probability'</span>) <span class="sc">+</span></span>
<span id="cb63-9"><a href="#cb63-9" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_ribbon</span>(<span class="fu">aes</span>(<span class="at">ymin =</span> low, <span class="at">ymax =</span> high, <span class="at">fill =</span> sex), <span class="at">alpha =</span> <span class="fl">0.2</span>, <span class="at">show.legend =</span> F) <span class="sc">+</span></span>
<span id="cb63-10"><a href="#cb63-10" aria-hidden="true" tabindex="-1"></a>  <span class="fu">theme_bw</span>(<span class="at">base_size =</span> <span class="dv">14</span>) <span class="sc">+</span> </span>
<span id="cb63-11"><a href="#cb63-11" aria-hidden="true" tabindex="-1"></a>  <span class="fu">scale_color_manual</span>(<span class="at">values =</span> my_colors) <span class="sc">+</span></span>
<span id="cb63-12"><a href="#cb63-12" aria-hidden="true" tabindex="-1"></a>  <span class="fu">scale_fill_manual</span>(<span class="at">values =</span> my_colors)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<div class="quarto-figure quarto-figure-center">
<figure class="figure">
<p><img src="bart_post_files/figure-html/surv-pdp-1.png" class="img-fluid figure-img" width="672"></p>
<figcaption class="figure-caption">Friedman’s partial dependence function with 95% prediction intervals: males vs females. Females show on average larger survival estimates compared to men, across all time points. Overlapping shaded area represents men and women that have similar survival characteristics.</figcaption>
</figure>
</div>
</div>
</div>
</section>
<section id="references" class="level2 unnumbered">


</section>

<div id="quarto-appendix" class="default"><section class="quarto-appendix-contents" role="doc-bibliography"><h2 class="anchored quarto-appendix-heading">References</h2><div id="refs" class="references csl-bib-body hanging-indent" role="list">
<div id="ref-Bonato2011" class="csl-entry" role="listitem">
Bonato, Vinicius, Veerabhadran Baladandayuthapani, Bradley M. Broom, Erik P. Sulman, Kenneth D. Aldape, and Kim Anh Do. 2011. <span>“<span class="nocase">Bayesian ensemble methods for survival prediction in gene expression data</span>.”</span> <em>Bioinformatics</em> 27 (3): 359–67. <a href="https://doi.org/10.1093/BIOINFORMATICS/BTQ660">https://doi.org/10.1093/BIOINFORMATICS/BTQ660</a>.
</div>
<div id="ref-Chipman2010" class="csl-entry" role="listitem">
Chipman, Hugh A, Edward I George, and Robert E McCulloch. 2010. <span>“<span>BART: BAYESIAN ADDITIVE REGRESSION TREES</span>.”</span> <em>The Annals of Applied Statistics</em> 4 (1): 266–98. <a href="http://www.jstor.org/stable/27801587">http://www.jstor.org/stable/27801587</a>.
</div>
<div id="ref-Friedman2001" class="csl-entry" role="listitem">
Friedman, Jerome H. 2001. <span>“<span class="nocase">Greedy function approximation: a gradient boosting machine</span>.”</span> <em>Annals of Statistics</em>, 1189–1232. <a href="https://doi.org/10.1214/aos/1013203451">https://doi.org/10.1214/aos/1013203451</a>.
</div>
<div id="ref-Sonabend2022" class="csl-entry" role="listitem">
Sonabend, Raphael, Andreas Bender, and Sebastian Vollmer. 2022. <span>“<span class="nocase">Avoiding C-hacking when evaluating survival distribution predictions with discrimination measures</span>.”</span> Edited by Zhiyong Lu. <em>Bioinformatics</em>, July. <a href="https://doi.org/10.1093/BIOINFORMATICS/BTAC451">https://doi.org/10.1093/BIOINFORMATICS/BTAC451</a>.
</div>
<div id="ref-Sonabend2022a" class="csl-entry" role="listitem">
Sonabend, Raphael, Florian Pfisterer, Alan Mishler, Moritz Schauer, Lukas Burk, Sumantrak Mukherjee, and Sebastian Vollmer. 2022. <span>“<span class="nocase">Flexible Group Fairness Metrics for Survival Analysis</span>,”</span> May. <a href="https://doi.org/10.48550/arxiv.2206.03256">https://doi.org/10.48550/arxiv.2206.03256</a>.
</div>
<div id="ref-Sparapani2016" class="csl-entry" role="listitem">
Sparapani, Rodney A., Brent R. Logan, Robert E. McCulloch, and Purushottam W. Laud. 2016. <span>“<span class="nocase">Nonparametric survival analysis using Bayesian Additive Regression Trees (BART)</span>.”</span> <em>Statistics in Medicine</em> 35 (16): 2741–53. <a href="https://doi.org/10.1002/SIM.6893">https://doi.org/10.1002/SIM.6893</a>.
</div>
<div id="ref-Sparapani2021" class="csl-entry" role="listitem">
Sparapani, Rodney, Charles Spanbauer, and Robert McCulloch. 2021. <span>“<span class="nocase">Nonparametric Machine Learning and Efficient Computation with Bayesian Additive Regression Trees: The BART R Package</span>.”</span> <em>Journal of Statistical Software</em> 97 (1): 1–66. <a href="https://doi.org/10.18637/JSS.V097.I01">https://doi.org/10.18637/JSS.V097.I01</a>.
</div>
</div></section></div></main>
<!-- /main column -->
<script id="quarto-html-after-body" type="application/javascript">
window.document.addEventListener("DOMContentLoaded", function (event) {
  const toggleBodyColorMode = (bsSheetEl) => {
    const mode = bsSheetEl.getAttribute("data-mode");
    const bodyEl = window.document.querySelector("body");
    if (mode === "dark") {
      bodyEl.classList.add("quarto-dark");
      bodyEl.classList.remove("quarto-light");
    } else {
      bodyEl.classList.add("quarto-light");
      bodyEl.classList.remove("quarto-dark");
    }
  }
  const toggleBodyColorPrimary = () => {
    const bsSheetEl = window.document.querySelector("link#quarto-bootstrap");
    if (bsSheetEl) {
      toggleBodyColorMode(bsSheetEl);
    }
  }
  toggleBodyColorPrimary();  
  const icon = "";
  const anchorJS = new window.AnchorJS();
  anchorJS.options = {
    placement: 'right',
    icon: icon
  };
  anchorJS.add('.anchored');
  const isCodeAnnotation = (el) => {
    for (const clz of el.classList) {
      if (clz.startsWith('code-annotation-')) {                     
        return true;
      }
    }
    return false;
  }
  const clipboard = new window.ClipboardJS('.code-copy-button', {
    text: function(trigger) {
      const codeEl = trigger.previousElementSibling.cloneNode(true);
      for (const childEl of codeEl.children) {
        if (isCodeAnnotation(childEl)) {
          childEl.remove();
        }
      }
      return codeEl.innerText;
    }
  });
  clipboard.on('success', function(e) {
    // button target
    const button = e.trigger;
    // don't keep focus
    button.blur();
    // flash "checked"
    button.classList.add('code-copy-button-checked');
    var currentTitle = button.getAttribute("title");
    button.setAttribute("title", "Copied!");
    let tooltip;
    if (window.bootstrap) {
      button.setAttribute("data-bs-toggle", "tooltip");
      button.setAttribute("data-bs-placement", "left");
      button.setAttribute("data-bs-title", "Copied!");
      tooltip = new bootstrap.Tooltip(button, 
        { trigger: "manual", 
          customClass: "code-copy-button-tooltip",
          offset: [0, -8]});
      tooltip.show();    
    }
    setTimeout(function() {
      if (tooltip) {
        tooltip.hide();
        button.removeAttribute("data-bs-title");
        button.removeAttribute("data-bs-toggle");
        button.removeAttribute("data-bs-placement");
      }
      button.setAttribute("title", currentTitle);
      button.classList.remove('code-copy-button-checked');
    }, 1000);
    // clear code selection
    e.clearSelection();
  });
  function tippyHover(el, contentFn) {
    const config = {
      allowHTML: true,
      content: contentFn,
      maxWidth: 500,
      delay: 100,
      arrow: false,
      appendTo: function(el) {
          return el.parentElement;
      },
      interactive: true,
      interactiveBorder: 10,
      theme: 'quarto',
      placement: 'bottom-start'
    };
    window.tippy(el, config); 
  }
  const noterefs = window.document.querySelectorAll('a[role="doc-noteref"]');
  for (var i=0; i<noterefs.length; i++) {
    const ref = noterefs[i];
    tippyHover(ref, function() {
      // use id or data attribute instead here
      let href = ref.getAttribute('data-footnote-href') || ref.getAttribute('href');
      try { href = new URL(href).hash; } catch {}
      const id = href.replace(/^#\/?/, "");
      const note = window.document.getElementById(id);
      return note.innerHTML;
    });
  }
      let selectedAnnoteEl;
      const selectorForAnnotation = ( cell, annotation) => {
        let cellAttr = 'data-code-cell="' + cell + '"';
        let lineAttr = 'data-code-annotation="' +  annotation + '"';
        const selector = 'span[' + cellAttr + '][' + lineAttr + ']';
        return selector;
      }
      const selectCodeLines = (annoteEl) => {
        const doc = window.document;
        const targetCell = annoteEl.getAttribute("data-target-cell");
        const targetAnnotation = annoteEl.getAttribute("data-target-annotation");
        const annoteSpan = window.document.querySelector(selectorForAnnotation(targetCell, targetAnnotation));
        const lines = annoteSpan.getAttribute("data-code-lines").split(",");
        const lineIds = lines.map((line) => {
          return targetCell + "-" + line;
        })
        let top = null;
        let height = null;
        let parent = null;
        if (lineIds.length > 0) {
            //compute the position of the single el (top and bottom and make a div)
            const el = window.document.getElementById(lineIds[0]);
            top = el.offsetTop;
            height = el.offsetHeight;
            parent = el.parentElement.parentElement;
          if (lineIds.length > 1) {
            const lastEl = window.document.getElementById(lineIds[lineIds.length - 1]);
            const bottom = lastEl.offsetTop + lastEl.offsetHeight;
            height = bottom - top;
          }
          if (top !== null && height !== null && parent !== null) {
            // cook up a div (if necessary) and position it 
            let div = window.document.getElementById("code-annotation-line-highlight");
            if (div === null) {
              div = window.document.createElement("div");
              div.setAttribute("id", "code-annotation-line-highlight");
              div.style.position = 'absolute';
              parent.appendChild(div);
            }
            div.style.top = top - 2 + "px";
            div.style.height = height + 4 + "px";
            let gutterDiv = window.document.getElementById("code-annotation-line-highlight-gutter");
            if (gutterDiv === null) {
              gutterDiv = window.document.createElement("div");
              gutterDiv.setAttribute("id", "code-annotation-line-highlight-gutter");
              gutterDiv.style.position = 'absolute';
              const codeCell = window.document.getElementById(targetCell);
              const gutter = codeCell.querySelector('.code-annotation-gutter');
              gutter.appendChild(gutterDiv);
            }
            gutterDiv.style.top = top - 2 + "px";
            gutterDiv.style.height = height + 4 + "px";
          }
          selectedAnnoteEl = annoteEl;
        }
      };
      const unselectCodeLines = () => {
        const elementsIds = ["code-annotation-line-highlight", "code-annotation-line-highlight-gutter"];
        elementsIds.forEach((elId) => {
          const div = window.document.getElementById(elId);
          if (div) {
            div.remove();
          }
        });
        selectedAnnoteEl = undefined;
      };
      // Attach click handler to the DT
      const annoteDls = window.document.querySelectorAll('dt[data-target-cell]');
      for (const annoteDlNode of annoteDls) {
        annoteDlNode.addEventListener('click', (event) => {
          const clickedEl = event.target;
          if (clickedEl !== selectedAnnoteEl) {
            unselectCodeLines();
            const activeEl = window.document.querySelector('dt[data-target-cell].code-annotation-active');
            if (activeEl) {
              activeEl.classList.remove('code-annotation-active');
            }
            selectCodeLines(clickedEl);
            clickedEl.classList.add('code-annotation-active');
          } else {
            // Unselect the line
            unselectCodeLines();
            clickedEl.classList.remove('code-annotation-active');
          }
        });
      }
  const findCites = (el) => {
    const parentEl = el.parentElement;
    if (parentEl) {
      const cites = parentEl.dataset.cites;
      if (cites) {
        return {
          el,
          cites: cites.split(' ')
        };
      } else {
        return findCites(el.parentElement)
      }
    } else {
      return undefined;
    }
  };
  var bibliorefs = window.document.querySelectorAll('a[role="doc-biblioref"]');
  for (var i=0; i<bibliorefs.length; i++) {
    const ref = bibliorefs[i];
    const citeInfo = findCites(ref);
    if (citeInfo) {
      tippyHover(citeInfo.el, function() {
        var popup = window.document.createElement('div');
        citeInfo.cites.forEach(function(cite) {
          var citeDiv = window.document.createElement('div');
          citeDiv.classList.add('hanging-indent');
          citeDiv.classList.add('csl-entry');
          var biblioDiv = window.document.getElementById('ref-' + cite);
          if (biblioDiv) {
            citeDiv.innerHTML = biblioDiv.innerHTML;
          }
          popup.appendChild(citeDiv);
        });
        return popup.innerHTML;
      });
    }
  }
});
</script>
</div> <!-- /content -->



</body></html>